### PR TITLE
IGNITE-15864 CPP Thin: Asynchronous network events handling

### PR DESCRIPTION
**Pull request by isapego**
Changes mainly for C++ thin client:
1. Introduced new network library class `AsyncClientPool`, which implemented using IOCP on Windows and epolls on Linux.
2. Adjusted C++ thin client logic accordingly. No async operations were added though -- this is a task for another ticket.
3. OpenSSL layer is now platform independent and introduced as `SecureDataFilter`.
4. Some other logic that is common for Linux and Windows moved to platform-independent classes, e.g. `CodecDataFilter`, `DataFilterAdapter`, `AsyncClientPoolAdapter` and so on.

### The Contribution Checklist
- [x] There is a single JIRA ticket related to the pull request. 
- [x] The web-link to the pull request is attached to the JIRA ticket.
- [x] The JIRA ticket has the _Patch Available_ state.
- [x] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [x] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [x] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [x] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/platforms/cpp/network/os/linux/src/network/linux_async_client.cpp [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (ptupitsyn) [inline modules/platforms/cpp/network/os/linux/src/network/linux_async_client.cpp:257]: Unnecessary blank lines.

File: modules/platforms/cpp/network/os/linux/src/network/linux_async_client_pool.cpp [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+            std::map<uint64_t, SP_LinuxAsyncClient>::const_iterator it = clientIdMap.find(id);
+            if (it == clientIdMap.end())
+                return SP_LinuxAsyncClient();
+
+            return it->second;
+        }
+    }
+}
+
+
```
Comment (ptupitsyn) [inline modules/platforms/cpp/network/os/linux/src/network/linux_async_client_pool.cpp:207]: Extra blank lines.

**Review by ptupitsyn — APPROVED** [review]

File: modules/platforms/cpp/network/os/linux/src/network/linux_async_client.cpp (hunk 1 above)
Comment (isapego) [reply modules/platforms/cpp/network/os/linux/src/network/linux_async_client.cpp:257]: Thanks!

File: modules/platforms/cpp/network/os/linux/src/network/linux_async_client_pool.cpp (hunk 2 above)
Comment (isapego) [reply modules/platforms/cpp/network/os/linux/src/network/linux_async_client_pool.cpp:207]: Removed.

File: modules/platforms/cpp/thin-client-test/src/sql_fields_query_test.cpp [hunk 3]
```diff
@@ -340,12 +340,12 @@ BOOST_AUTO_TEST_CASE(Select10000Values)
 
     for (int64_t i = 0; i < num; ++i)
     {
-        BOOST_CHECK(cursor.HasNext());
+//        BOOST_CHECK(cursor.HasNext());
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/thin-client-test/src/sql_fields_query_test.cpp:343]: No, it should be uncommented actually 😄

File: modules/platforms/cpp/common/include/ignite/common/concurrent.h [hunk 4]
```diff
@@ -20,6 +20,7 @@
 
 #include <cassert>
 #include <utility>
+#include <iostream>
```
Comment (ivandasch) [inline modules/platforms/cpp/common/include/ignite/common/concurrent.h:23]: Is `iostream` needed here?

File: modules/platforms/cpp/common/include/ignite/future.h [hunk 5]
```diff
@@ -279,6 +279,130 @@ namespace ignite
...
+
+        /**
+         * Wait for value to be set.
+         * Active thread will be blocked until value or error will be set.
+         */
+        void Wait() const
+        {
+            const common::SharedState<SP_ValueType>* state0 = state.Get();
+
+            assert(state0 != 0);
```
Comment (ivandasch) [inline modules/platforms/cpp/common/include/ignite/future.h:330]: MAY be check for not  `NULL`?

File: modules/platforms/cpp/common/os/linux/src/common/concurrent_os.cpp [hunk 6]
```diff
@@ -203,6 +203,39 @@ namespace ignite
...
+                Thread* self = static_cast<Thread*>(arg);
+
+                self->Run();
+
+                return 0;
+            }
+
+            void Thread::Start()
+            {
+                int res = pthread_create(&thread, 0, Thread::ThreadRoutine, this);
```
Comment (ivandasch) [inline modules/platforms/cpp/common/os/linux/src/common/concurrent_os.cpp:229]: May be use `NULL` here?

File: modules/platforms/cpp/network/include/ignite/network/data_buffer.h [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                // No-op.
+            }
+
+            /**
+             * Copy data from buffer to specified place in memory.
+             *
+             * @param dst Destination in memory.
+             * @param size Number of bytes to copy.
+             */
+            void GetData(int8_t* dst, int32_t size);
```
Comment (ivandasch) [inline modules/platforms/cpp/network/include/ignite/network/data_buffer.h:69]: May be rename to `CopyTo`? It'c a little bit embarrassing naming, IMHO.
And is it necessary to use signed size here?

**Review by ivandasch — APPROVED** [review]

File: modules/platforms/cpp/common/include/ignite/common/concurrent.h (hunk 4 above)
Comment (isapego) [reply modules/platforms/cpp/common/include/ignite/common/concurrent.h:23]: No :)

File: modules/platforms/cpp/common/include/ignite/future.h [hunk 8]
```diff
@@ -279,6 +279,130 @@ namespace ignite
...
+        }
+
+        /**
+         * Assignment operator.
+         *
+         * @param other Other instance.
+         * @return *this.
+         */
+        Future& operator=(const Future<SP_ValueType>& other)
+        {
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/common/include/ignite/future.h:316]: Fixed

File: modules/platforms/cpp/common/include/ignite/future.h (hunk 5 above)
Comment (isapego) [reply modules/platforms/cpp/common/include/ignite/future.h:330]: In C++ standard there is no `NULL` - this is C constant, in C++03 and below simple 0 constant is used, when not used with C API.

File: modules/platforms/cpp/common/os/linux/src/common/concurrent_os.cpp (hunk 6 above)
Comment (isapego) [reply modules/platforms/cpp/common/os/linux/src/common/concurrent_os.cpp:229]: Makes sense in this place

File: modules/platforms/cpp/network/include/ignite/network/data_buffer.h (hunk 7 above)
Comment (isapego) [reply modules/platforms/cpp/network/include/ignite/network/data_buffer.h:69]: Agree that naming is too general :). But I don't like `CopyTo` as well - to me anything with `Copy...` in it's name should be `const` and should not change internal state of the object. Maybe `Consume`?

File: modules/platforms/cpp/network/os/linux/src/network/linux_async_worker_thread.cpp [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+                epoll_event& currentEvent = events[i];
+                LinuxAsyncClient* client = static_cast<LinuxAsyncClient*>(currentEvent.data.ptr);
+                if (!client)
+                    continue;
+
+                if (client == currentClient.Get())
+                {
+                    if (currentEvent.events & (EPOLLRDHUP | EPOLLERR))
+                    {
+                        HandleConnectionFailed("Can not establsih connection");
```
Comment (ivandasch) [inline modules/platforms/cpp/network/os/linux/src/network/linux_async_worker_thread.cpp:234]: Typo -- `establish`

Comment (isapego) [reply modules/platforms/cpp/network/os/linux/src/network/linux_async_worker_thread.cpp]: Fixed

**Pull request closed without merging**